src/services/localSSHService.ts

```diff
@@ -0,0 +1,102 @@
+    initialized: Promise<void>;
+}
+
+export class LocalSSHService extends Disposable implements ILocalSSHService {
+    public isSupportLocalSSH: boolean = false;
+    public initialized: Promise<void>;
+    constructor(
+        private readonly context: vscode.ExtensionContext,
+        private readonly hostService: IHostService,
+        private readonly logService: ILogService
+    ) {
+        super();
+
+        this.initialized = this.initialize();
+
+        this._register(vscode.workspace.onDidChangeConfiguration(async e => {
+            if (
+                e.affectsConfiguration('gitpod.lsshExtensionIpcPort') ||
+                e.affectsConfiguration('gitpod.host') ||
+                e.affectsConfiguration('remote.SSH.configFile')
+            ) {
+                if (e.affectsConfiguration('[javascript]') && e.affectsConfiguration('[markdown]')) {
+                    // Seems onDidChangeConfiguration fires many times while resolving the remote (once with all settings),
+                    // and because now we active the extension earlier with onResolveRemoteAuthority we get this false positive
+                    // event, so ignore it if more settings are affected at the same time.
+                    return;
+                }
+                this.initialize();
+            }
+        }));
+    }
+
+    private async initialize() {
+        if (this.context.extensionMode !== vscode.ExtensionMode.Production) {
+            // TODO: add webpack config for client.js in development, for now copy manually
+            this.isSupportLocalSSH = true;
+            return;
+        }
+
+        try {
+            const locations = await this.copyProxyScript();
+            await this.configureSettings(locations);
+            this.isSupportLocalSSH = true;
+        } catch (e) {
+            this.logService.error(e, 'failed to copy local ssh client.js');
+            this.isSupportLocalSSH = false;
+        }
+    }
+
+    private async configureSettings({ proxyScript, launcher }: { proxyScript: string; launcher: string }) {
+        const extIpcPort = Configuration.getLocalSshExtensionIpcPort();
+        const hostConfig = this.getHostSSHConfig(this.hostService.gitpodHost, launcher, proxyScript, extIpcPort);
+        await SSHConfiguration.ensureIncludeGitpodSSHConfig();
+        const gitpodConfig = await SSHConfiguration.loadGitpodSSHConfig();
+        gitpodConfig.addHostConfiguration(hostConfig);
+        await SSHConfiguration.saveGitpodSSHConfig(gitpodConfig);
+    }
+
+    private getHostSSHConfig(host: string, launcher: string, proxyScript: string, extIpcPort: number) {
+        return {
+            Host: '*.' + getLocalSSHDomain(host),
+            StrictHostKeyChecking: 'no',
+            ProxyCommand: `"${launcher}" "${process.execPath}" "${proxyScript}" --ms-enable-electron-run-as-node %h ${extIpcPort} "${Configuration.getLocalSSHLogPath()}"`
+        };
+    }
+
+    private async copyProxyScript() {
```

Comment:
again just to my information:
- why do we need to copy cannot we execute from extensions dir?
- how upgrade is happening? we copy on each window?

Reply:
- extension dir is not stable it changes with every update as version is part of the folder name
- yep, maybe we can have a control file to keep track we latest version, but not sure if it's worth it, we need to check if all files exist either way to ensure no edge cases of somehow one of them missing for any reason, I'm also trying to reduce bundle size of proxy script so copy is fast right now is 1.4MB :grimacing: